cnap/core/filedb.py

```diff
@@ -0,0 +1,70 @@
+"""
+This module provides an object-oriented design for file database to load video file.
+It implements an abstract base class `FileDatabase` and a concrete filedatabase class
+`LocalFileDatabase`.
+
+`FileDatabase` serves as a blueprint for custom filedatabase implementations, while
+`LocalFileDatabase` provide an implementation for local filedatabase to get video 
+file locally.
+
+These classes can be easily extended or modified to accommodate new filedatabase type.
+"""
+
+import os
+import logging
+from abc import ABC, abstractmethod
+
+LOG = logging.getLogger(__name__)
+
+class FileDatabase(ABC):
+    """
+    Abstract base class for creating custom filedatabase implementations.
+    """
+
+    def __init__(self):
+        pass
+
+    @abstractmethod
+    def get_file(self, filename: str) -> str:
+        """
+        Get/Download file from file database and return the local path for file.
+
+        Args:
+            filename: The name of file to get/Download.
+
+        Returns:
+            str: The local path for file.
+        """
+        raise NotImplementedError("Subclasses should implement get_file() method.")
+
+class LocalFileDatabase(FileDatabase):
+    """
+    Local filedata base implementation for getting video file locally.
+    """
+
+    def __init__(self, root_dir: str):
+        self._root_dir = os.path.abspath(root_dir)
+
+    def get_file(self, filename: str) -> str:
+        """
+        Get the local path for file.
+
+        Args:
+            filename: The name of file to get.
+
+        Returns:
+            str: The local path for file.
+
+        Raises:
+            RuntimeError: if the root_dir is invalid or filed to find the given file.
+        """
+        if not os.path.exists(self._root_dir):
+            LOG.error("Invalid root directory for local file database.")
+            raise RuntimeError("Invalid root directory for local file database.")
```

Comment:
Maybe `NotADirectoryError` or something more specified to this case, please help check

Reply:
How about also use `FileNotFoundError`? Refer to https://docs.python.org/3/library/exceptions.html

exception NotADirectoryError
Raised when a directory operation (such as [os.listdir()](https://docs.python.org/3/library/os.html#os.listdir)) is requested on something which is not a directory. On most POSIX platforms, it may also be raised if an operation attempts to open or traverse a non-directory file as if it were a directory. Corresponds to errno [ENOTDIR](https://docs.python.org/3/library/errno.html#errno.ENOTDIR).

exception FileNotFoundError
Raised when a file **or directory** is requested but doesn’t exist. Corresponds to errno [ENOENT](https://docs.python.org/3/library/errno.html#errno.ENOENT).

Reply:
It makes sense.